### DISCUSSION TITLE
`'expandtab'`-agnostic indentation

### DISCUSSION THREAD

**Discussion by kraftwerk28** [L3MON4D3/LuaSnip#337] [Q&A] [answered]
Is it possible to create a multiline snippet that would have an indentation based on `et`/`ts`/`sw` options in the buffer?
For example:
```lua
s(
  "ie=",
  fmt([[
    if err := {}; err != nil {{
    {}{}
    }}
  ]], { i(1), agnostic_indent_node(), i(0) })
)
```

**Comment by kraftwerk28** [comment]
The `agnostic_indent_node` should insert a single tab if `noexpandtab` is set, 2 spaces if `expandtab, shiftwidth=2` etc.

**Reply by bew** [reply]
idea: In Ultisnips, a tab in the snippet would do exactly that (match the tab/space config of the buffer), maybe LuaSnip could copy this behavior?

**Reply by leiserfg** [reply]
The machinery is in place, maybe is just that the `fmt` helper is not using it.

**Reply by kraftwerk28** [reply]
> idea: In Ultisnips, a tab in the snippet would do exactly that (match the tab/space config of the buffer), maybe LuaSnip could copy this behavior?

Hm, I just realized that LuaSnip also has this behavior, i.e. `t("\t")` matches et/sw/ts. Thank you.

**Comment by leiserfg** [comment, marked as answer]
Try:
```lua
    s(
      "ie=",
      fmt([[
        if err := {}; err != nil {{
        {}{}
        }}
      ]], { i(1), t"\t", i(0) })
    ),
```

**Comment by cskeeters** [comment]
[Late Nov 2024](https://github.com/L3MON4D3/LuaSnip/commit/8f284020c993a06a998e8f5d4c6f19325a350402), `indent_string` was added as an option to `fmt`.   When provided, `fmt` will identify indentation in lines of `format` argument by finding instances of `input_string` at the start of the lines.  This happens after leading space common to all lines are removed (see `dedent`).  Using this method, `fmt` will replace the indentation found in the format string with indentation appropriate for the buffer.  This method does not require node for indentation, which makes it more elegant IMHO.

```
s(
    "ie=",
    fmt([[
        if err := {}; err != nil {{
          {}
        }}
      ]],
      { i(1), i(0) },
      { indent_string = "  " }
    )
)
```
